### ISSUE TITLE
Separate constraints file download breaks Docker image builds (IDFGH-7722)

### ISSUE THREAD

**Issue by ftab** [espressif/esp-idf#9263]
## Environment

- Development Kit:      all of em
- Module or chip used:  all of em
- IDF version (run ``git describe --tags`` to find it): v5.0-dev-1426-ge899edd793
- Build System:        idf.py
- Compiler version (run ``xtensa-esp32-elf-gcc --version`` to find it): xtensa-esp32-elf-gcc (crosstool-NG esp-2021r2) 8.4.0
- Operating System:     Linux
- Using an IDE?: no
- Power Supply:         Unlimited power

## Problem Description

The commit here https://github.com/espressif/esp-idf/commit/b28d7e6850936a9e5bbf542fa1d55adf6f9ae2f3 caused esp-idf to start downloading a constraints file from the internet on every install / export.

This eventually broke my Docker build container which is a custom build of esp-adf that works around some odd issues and contains a specific checkout of esp-idf at the aforementioned e899edd793. (the build fails on recent masters but at the time of e899edd793 a bug was fixed that hadn't been backported to 4.4 at the time I made the container, so it was deemed necessary to venture into unstable territory)

At the time I built the container, everything was working. At some point between then and now, the constraints file changed. (somewhere between June 6-9 by the looks of where our job logs started all failing)

Now, all builds fail because the new constraints file is not satisfied. Rebuilding the Docker container should work, but that means the Dockerfile is not reproducible. It depends on a specific point in time and not just a specific git version.

I see a comment in the commit:

    # We need to constrain package dependencies as well. So all installed packages need to be checked.
    # For example package A requires package B. We have only A in our requirements. But the newest version of B could
    # broke at some time and in that case we add a constraint for B (on the server) but don't have to update the
    # requirement file (in the ESP-IDF repo).

I appreciate the desire to not have to do a whole commit-release cycle to fix something like that. Ironically enough, the fact that this file changed on the server **is the only thing that actually broke something**, at least so far, for me. Granted, this is an unreleased, unstable branch as of yet. But even so, if it is discovered that a newer version of a sub-dependency package does break something, I think it's best for it to be committed in the repo so that the same git checkout will build the same build environment every time. Ideally, dependencies are locked in at some way. I am not sure what sort of blackbox dependency hell auto update torture prompted this change, but the cure feels worse than the disease. Unfortunately the alternative does mean, in the event it happens, new checkout/installs of old versions that lack the necessary upper limits on dependency versions will of course break.

### Expected Behavior

Reproducible builds

### Actual Behavior

Build tool downloads an external file which is not a part of the repo which can break things when it changes outside of version control

### Steps to reproduce

1. go back in time to before https://dl.espressif.com/dl/esp-idf/espidf.constraints.v5.0.txt required idf-component-manager~=1.1
2. make a Docker image for an esp-idf build container based on any checkout from b28d7e6850936a9e5bbf542fa1d55adf6f9ae2f3 or newer. i.e. https://github.com/espressif/esp-idf/blob/master/tools/docker/Dockerfile with a specific `IDF_CHECKOUT_REF` 
3. go back forward in time
4. try to build using the Docker image produced in step 2

## CI Job Log

```
...
Downloading espidf.constraints.v5.0.txt to /root/.espressif/espidf.constraints.v5.0.txt.tmp
Done
The following Python requirements are not satisfied:
idf-component-manager~=1.1
To install the missing packages, please run "install.sh"
Diagnostic information:
    IDF_PYTHON_ENV_PATH: /opt/esp/python_env/idf5.0_py3.8_env
    Python interpreter used: /opt/esp/python_env/idf5.0_py3.8_env/bin/python
Constraint file: /root/.espressif/espidf.constraints.v5.0.txt
Requirement files:
 - /opt/esp/adf/esp-idf/requirements.core.txt
Python being checked: /opt/esp/python_env/idf5.0_py3.8_env/bin/python
Cleaning up project directory and file based variables
00:01
ERROR: Job failed: exit code 1
```

**Comment by ftab** [comment]
Perhaps something like pip-tools can help. It bakes the requirements.txt file similar to how a Gemfile.lock works for ruby gems if I understand correctly

https://github.com/jazzband/pip-tools

Or just an option to opt into the newer constraints file

**Comment by dobairoland** [comment]
First of all, we are sorry for the inconvenience. The master branch currently is the work-in-progress v5.0 which is a major release and will contain breaking changes in comparison with v4.4.x. You cannot except to get the same results for all commits between the top of the v4.4 and the master branch. Once v5.0 will be released, the constraint file will freeze and will be modified only to solve potential bugs.

Back to your problem. I can help to work around the issue. I assume that you want to keep using `install.sh` and `export.sh`. (At the lower level script there are options to avoid the constraint file.) For the use case with `install.sh` and `export.sh` you can use the following.

You can modify your `/root/.espressif/espidf.constraints.v5.0.txt` file with requirements which suits your needs (compatibility with older master commits).  After that, you can add `touch /root/.espressif/espidf.constraints.v5.0.txt` to the beginning of your docker file before calling `install.sh` or `export.sh`. That will avoid the download and overwite with a newer version from the server. 

I hope this helps.

**Comment by ftab** [comment]
Thanks for your attention to this issue. Sorry to trouble you.

I know that 5.0 is not released yet, but when it is, the same problem will apply. Everybody building with containers will run into it (built themselves or downloaded from the espressif docker hub)--eventually the image they're using will become out of sync with the constraints file on the server and builds will fail.

I should mention I did expect to be traveling some bumpy roads going off of the stable branch, and I usually don't bother anyone about what I run into since it's in development, and that's how it goes. This is a little different than the usual bleeding-edge branch excitement, if it's how it's going to be done once it is stable.

> Once v5.0 will be released, the constraint file will freeze and will be modified only to solve potential bugs.

Will it freeze, or will it be modified to solve potential bugs? Those are two different things.
* If the constraints file for that version ever does change (even to solve potential bugs), I suspect the build will still fail, the installed versions will be different than the ones stated in the constraints file downloaded at the time of building the image
* If the file will not change once released, why not commit to version control along with the rest of it?

The constraints format seems prone to causing confusion. It certainly confused me when I ran into it this month. Maybe that's just the covid that's taking a while to wear off.

I still think if you could try [pip-tools](https://github.com/jazzband/pip-tools) it might be a better way. They solve this for you by pinning the versions. You don't automatically get new features or fixes from future dependency updates, but that's the only downside. The major tradeoff is at least things are much less likely to break outside of your control.

e.g. if I move requirements.core.txt to requirements.core.in and then `pip-compile requirements.core.in -o requirements.core.txt --generate-hashes --allow-unsafe` it generates the below requirements.core.txt, as of right now (latest master)

```
#
# This file is autogenerated by pip-compile with python 3.8
# To update, run:
#
#    pip-compile --allow-unsafe --generate-hashes --output-file=requirements.core.txt requirements.core.in
#
bitstring==3.1.9 \
    --hash=sha256:0de167daa6a00c9386255a7cac931b45e6e24e0ad7ea64f1f92a64ac23ad4578 \
    --hash=sha256:a5848a3f63111785224dca8bb4c0a75b62ecdef56a042c8d6be74b16f7e860e7 \
    --hash=sha256:e3e340e58900a948787a05e8c08772f1ccbe133f6f41fe3f0fa19a18a22bbf4f
    # via esptool
certifi==2022.6.15 \
    --hash=sha256:84c85a9078b11105f04f3036a9482ae10e4621616db313fe045dd24743a0820d \
    --hash=sha256:fe86415d55e84719d75f8b69414f6438ac3547d2078ab91b67e779ef69378412
    # via requests
cffi==1.15.1 \
    --hash=sha256:00a9ed42e88df81ffae7a8ab6d9356b371399b91dbdf0c3cb1e84c03a13aceb5 \
    --hash=sha256:03425bdae262c76aad70202debd780501fabeaca237cdfddc008987c0e0f59ef \
    --hash=sha256:04ed324bda3cda42b9b695d51bb7d54b680b9719cfab04227cdd1e04e5de3104 \
    --hash=sha256:0e2642fe3142e4cc4af0799748233ad6da94c62a8bec3a6648bf8ee68b1c7426 \
    --hash=sha256:173379135477dc8cac4bc58f45db08ab45d228b3363adb7af79436135d028405 \
    --hash=sha256:198caafb44239b60e252492445da556afafc7d1e3ab7a1fb3f0584ef6d742375 \
    --hash=sha256:1e74c6b51a9ed6589199c787bf5f9875612ca4a8a0785fb2d4a84429badaf22a \
    --hash=sha256:2012c72d854c2d03e45d06ae57f40d78e5770d252f195b93f581acf3ba44496e \
    --hash=sha256:21157295583fe8943475029ed5abdcf71eb3911894724e360acff1d61c1d54bc \
    --hash=sha256:2470043b93ff09bf8fb1d46d1cb756ce6132c54826661a32d4e4d132e1977adf \
    --hash=sha256:285d29981935eb726a4399badae8f0ffdff4f5050eaa6d0cfc3f64b857b77185 \
    --hash=sha256:30d78fbc8ebf9c92c9b7823ee18eb92f2e6ef79b45ac84db507f52fbe3ec4497 \
    --hash=sha256:320dab6e7cb2eacdf0e658569d2575c4dad258c0fcc794f46215e1e39f90f2c3 \
    --hash=sha256:33ab79603146aace82c2427da5ca6e58f2b3f2fb5da893ceac0c42218a40be35 \
    --hash=sha256:3548db281cd7d2561c9ad9984681c95f7b0e38881201e157833a2342c30d5e8c \
    --hash=sha256:3799aecf2e17cf585d977b780ce79ff0dc9b78d799fc694221ce814c2c19db83 \
    --hash=sha256:39d39875251ca8f612b6f33e6b1195af86d1b3e60086068be9cc053aa4376e21 \
    --hash=sha256:3b926aa83d1edb5aa5b427b4053dc420ec295a08e40911296b9eb1b6170f6cca \
    --hash=sha256:3bcde07039e586f91b45c88f8583ea7cf7a0770df3a1649627bf598332cb6984 \
    --hash=sha256:3d08afd128ddaa624a48cf2b859afef385b720bb4b43df214f85616922e6a5ac \
    --hash=sha256:3eb6971dcff08619f8d91607cfc726518b6fa2a9eba42856be181c6d0d9515fd \
    --hash=sha256:40f4774f5a9d4f5e344f31a32b5096977b5d48560c5592e2f3d2c4374bd543ee \
    --hash=sha256:4289fc34b2f5316fbb762d75362931e351941fa95fa18789191b33fc4cf9504a \
    --hash=sha256:470c103ae716238bbe698d67ad020e1db9d9dba34fa5a899b5e21577e6d52ed2 \
    --hash=sha256:4f2c9f67e9821cad2e5f480bc8d83b8742896f1242dba247911072d4fa94c192 \
    --hash=sha256:50a74364d85fd319352182ef59c5c790484a336f6db772c1a9231f1c3ed0cbd7 \
    --hash=sha256:54a2db7b78338edd780e7ef7f9f6c442500fb0d41a5a4ea24fff1c929d5af585 \
    --hash=sha256:5635bd9cb9731e6d4a1132a498dd34f764034a8ce60cef4f5319c0541159392f \
    --hash=sha256:59c0b02d0a6c384d453fece7566d1c7e6b7bae4fc5874ef2ef46d56776d61c9e \
    --hash=sha256:5d598b938678ebf3c67377cdd45e09d431369c3b1a5b331058c338e201f12b27 \
    --hash=sha256:5df2768244d19ab7f60546d0c7c63ce1581f7af8b5de3eb3004b9b6fc8a9f84b \
    --hash=sha256:5ef34d190326c3b1f822a5b7a45f6c4535e2f47ed06fec77d3d799c450b2651e \
    --hash=sha256:6975a3fac6bc83c4a65c9f9fcab9e47019a11d3d2cf7f3c0d03431bf145a941e \
    --hash=sha256:6c9a799e985904922a4d207a94eae35c78ebae90e128f0c4e521ce339396be9d \
    --hash=sha256:70df4e3b545a17496c9b3f41f5115e69a4f2e77e94e1d2a8e1070bc0c38c8a3c \
    --hash=sha256:7473e861101c9e72452f9bf8acb984947aa1661a7704553a9f6e4baa5ba64415 \
    --hash=sha256:8102eaf27e1e448db915d08afa8b41d6c7ca7a04b7d73af6514df10a3e74bd82 \
    --hash=sha256:87c450779d0914f2861b8526e035c5e6da0a3199d8f1add1a665e1cbc6fc6d02 \
    --hash=sha256:8b7ee99e510d7b66cdb6c593f21c043c248537a32e0bedf02e01e9553a172314 \
    --hash=sha256:91fc98adde3d7881af9b59ed0294046f3806221863722ba7d8d120c575314325 \
    --hash=sha256:94411f22c3985acaec6f83c6df553f2dbe17b698cc7f8ae751ff2237d96b9e3c \
    --hash=sha256:98d85c6a2bef81588d9227dde12db8a7f47f639f4a17c9ae08e773aa9c697bf3 \
    --hash=sha256:9ad5db27f9cabae298d151c85cf2bad1d359a1b9c686a275df03385758e2f914 \
    --hash=sha256:a0b71b1b8fbf2b96e41c4d990244165e2c9be83d54962a9a1d118fd8657d2045 \
    --hash=sha256:a0f100c8912c114ff53e1202d0078b425bee3649ae34d7b070e9697f93c5d52d \
    --hash=sha256:a591fe9e525846e4d154205572a029f653ada1a78b93697f3b5a8f1f2bc055b9 \
    --hash=sha256:a5c84c68147988265e60416b57fc83425a78058853509c1b0629c180094904a5 \
    --hash=sha256:a66d3508133af6e8548451b25058d5812812ec3798c886bf38ed24a98216fab2 \
    --hash=sha256:a8c4917bd7ad33e8eb21e9a5bbba979b49d9a97acb3a803092cbc1133e20343c \
    --hash=sha256:b3bbeb01c2b273cca1e1e0c5df57f12dce9a4dd331b4fa1635b8bec26350bde3 \
    --hash=sha256:cba9d6b9a7d64d4bd46167096fc9d2f835e25d7e4c121fb2ddfc6528fb0413b2 \
    --hash=sha256:cc4d65aeeaa04136a12677d3dd0b1c0c94dc43abac5860ab33cceb42b801c1e8 \
    --hash=sha256:ce4bcc037df4fc5e3d184794f27bdaab018943698f4ca31630bc7f84a7b69c6d \
    --hash=sha256:cec7d9412a9102bdc577382c3929b337320c4c4c4849f2c5cdd14d7368c5562d \
    --hash=sha256:d400bfb9a37b1351253cb402671cea7e89bdecc294e8016a707f6d1d8ac934f9 \
    --hash=sha256:d61f4695e6c866a23a21acab0509af1cdfd2c013cf256bbf5b6b5e2695827162 \
    --hash=sha256:db0fbb9c62743ce59a9ff687eb5f4afbe77e5e8403d6697f7446e5f609976f76 \
    --hash=sha256:dd86c085fae2efd48ac91dd7ccffcfc0571387fe1193d33b6394db7ef31fe2a4 \
    --hash=sha256:e00b098126fd45523dd056d2efba6c5a63b71ffe9f2bbe1a4fe1716e1d0c331e \
    --hash=sha256:e229a521186c75c8ad9490854fd8bbdd9a0c9aa3a524326b55be83b54d4e0ad9 \
    --hash=sha256:e263d77ee3dd201c3a142934a086a4450861778baaeeb45db4591ef65550b0a6 \
    --hash=sha256:ed9cb427ba5504c1dc15ede7d516b84757c3e3d7868ccc85121d9310d27eed0b \
    --hash=sha256:fa6693661a4c91757f4412306191b6dc88c1703f780c8234035eac011922bc01 \
    --hash=sha256:fcd131dd944808b5bdb38e6f5b53013c5aa4f334c5cad0c72742f6eba4b73db0
    # via cryptography
charset-normalizer==2.1.0 \
    --hash=sha256:5189b6f22b01957427f35b6a08d9a0bc45b46d3788ef5a92e978433c7a35f8a5 \
    --hash=sha256:575e708016ff3a5e3681541cb9d79312c416835686d054a23accb873b254f413
    # via requests
click==8.1.3 \
    --hash=sha256:7682dc8afb30297001674575ea00d1814d808d6a36af415a82bd481d37ba7b8e \
    --hash=sha256:bb4d8133cb15a609f44e8213d9b391b0809795062913b383c62be0ee95b1db48
    # via
    #   -r requirements.core.in
    #   idf-component-manager
colorama==0.4.5 \
    --hash=sha256:854bf444933e37f5824ae7bfc1e98d5bce2ebe4160d46b5edf346a89358e99da \
    --hash=sha256:e6c6b4334fc50988a639d9b98aa429a0b57da6e17b9a44f0451f930b6967b7a4
    # via idf-component-manager
construct==2.10.68 \
    --hash=sha256:7b2a3fd8e5f597a5aa1d614c3bd516fa065db01704c72a1efaaeec6ef23d8b45
    # via esp-coredump
contextlib2==21.6.0 \
    --hash=sha256:3fbdb64466afd23abaf6c977627b75b6139a5a3e8ce38405c5b413aed7a0471f \
    --hash=sha256:ab1e2bfe1d01d968e1b7e8d9023bc51ef3509bba217bb730cee3827e1ee82869
    # via
    #   idf-component-manager
    #   schema
cryptography==37.0.2 \
    --hash=sha256:093cb351031656d3ee2f4fa1be579a8c69c754cf874206be1d4cf3b542042804 \
    --hash=sha256:0cc20f655157d4cfc7bada909dc5cc228211b075ba8407c46467f63597c78178 \
    --hash=sha256:1b9362d34363f2c71b7853f6251219298124aa4cc2075ae2932e64c91a3e2717 \
    --hash=sha256:1f3bfbd611db5cb58ca82f3deb35e83af34bb8cf06043fa61500157d50a70982 \
    --hash=sha256:2bd1096476aaac820426239ab534b636c77d71af66c547b9ddcd76eb9c79e004 \
    --hash=sha256:31fe38d14d2e5f787e0aecef831457da6cec68e0bb09a35835b0b44ae8b988fe \
    --hash=sha256:3b8398b3d0efc420e777c40c16764d6870bcef2eb383df9c6dbb9ffe12c64452 \
    --hash=sha256:3c81599befb4d4f3d7648ed3217e00d21a9341a9a688ecdd615ff72ffbed7336 \
    --hash=sha256:419c57d7b63f5ec38b1199a9521d77d7d1754eb97827bbb773162073ccd8c8d4 \
    --hash=sha256:46f4c544f6557a2fefa7ac8ac7d1b17bf9b647bd20b16decc8fbcab7117fbc15 \
    --hash=sha256:471e0d70201c069f74c837983189949aa0d24bb2d751b57e26e3761f2f782b8d \
    --hash=sha256:59b281eab51e1b6b6afa525af2bd93c16d49358404f814fe2c2410058623928c \
    --hash=sha256:731c8abd27693323b348518ed0e0705713a36d79fdbd969ad968fbef0979a7e0 \
    --hash=sha256:95e590dd70642eb2079d280420a888190aa040ad20f19ec8c6e097e38aa29e06 \
    --hash=sha256:a68254dd88021f24a68b613d8c51d5c5e74d735878b9e32cc0adf19d1f10aaf9 \
    --hash=sha256:a7d5137e556cc0ea418dca6186deabe9129cee318618eb1ffecbd35bee55ddc1 \
    --hash=sha256:aeaba7b5e756ea52c8861c133c596afe93dd716cbcacae23b80bc238202dc023 \
    --hash=sha256:dc26bb134452081859aa21d4990474ddb7e863aa39e60d1592800a8865a702de \
    --hash=sha256:e53258e69874a306fcecb88b7534d61820db8a98655662a3dd2ec7f1afd9132f \
    --hash=sha256:ef15c2df7656763b4ff20a9bc4381d8352e6640cfeb95c2972c38ef508e75181 \
    --hash=sha256:f224ad253cc9cea7568f49077007d2263efa57396a2f2f78114066fd54b5c68e \
    --hash=sha256:f8ec91983e638a9bcd75b39f1396e5c0dc2330cbd9ce4accefe68717e6779e0a
    # via
    #   -r requirements.core.in
    #   esptool
ecdsa==0.17.0 \
    --hash=sha256:5cf31d5b33743abe0dfc28999036c849a69d548f994b535e527ee3cb7f3ef676 \
    --hash=sha256:b9f500bb439e4153d0330610f5d26baaf18d17b8ced1bc54410d189385ea68aa
    # via esptool
esp-coredump==1.2 \
    --hash=sha256:c5c043332ff5fd4668e4ad7e9d7004db6d136f60da3f65604a478de5e0b75d88
    # via -r requirements.core.in
esptool==4.1 \
    --hash=sha256:51a41655187780cb802a5827fab68a3db0d6206330560459c55fcf0e222f277f
    # via
    #   -r requirements.core.in
    #   esp-coredump
freertos-gdb==1.0.1 \
    --hash=sha256:5e0b5120528d025aee8a019e2b59f2e553845a301fc78b3d5f2c25bd23e19a08 \
    --hash=sha256:da940d6a3388b728e9b87919f0acbcedb48a3e8ef9d86576ddb3106937a06191
    # via -r requirements.core.in
future==0.18.2 \
    --hash=sha256:b1bead90b70cf6ec3f0710ae53a525360fa360d306a86583adc6bf83a4db537d
    # via
    #   -r requirements.core.in
    #   idf-component-manager
idf-component-manager==1.1.3 \
    --hash=sha256:5a9f236f259cebbd4b069bdac64b99ec53c68cf3216e59fcb80b62dfe72eda32 \
    --hash=sha256:c1cd9ef9621486cc2faa16d77d1b8b44ebe4bdb6d72b4628245722886a2b40f9
    # via -r requirements.core.in
idna==3.3 \
    --hash=sha256:84d9dd047ffa80596e0f246e2eab0b391788b0503584e8945f2368256d2735ff \
    --hash=sha256:9d643ff0a55b762d5cdb124b8eaa99c66322e2157b69160bc32796e824360e6d
    # via requests
kconfiglib==14.1.0 \
    --hash=sha256:bed2cc2216f538eca4255a83a4588d8823563cdd50114f86cf1a2674e602c93c \
    --hash=sha256:edcd35a20e7e138a9a9e96149027f805f785e818d2eae400b6fa8b0c8845114a
    # via -r requirements.core.in
pycparser==2.21 \
    --hash=sha256:8ee45429555515e1f6b185e78100aea234072576aa43ab53aefcae078162fca9 \
    --hash=sha256:e644fdec12f7872f86c58ff790da456218b10f863970249516d60a5eaca77206
    # via cffi
pyelftools==0.28 \
    --hash=sha256:1509d30756a14a8c9a6670778144fb9a04b311751ddac019dcc0c2f28da03ccc \
    --hash=sha256:53e5609cac016471d40bd88dc410cd90755942c25e58a61021cfdf7abdfeacff
    # via -r requirements.core.in
pygdbmi==0.10.0.2 \
    --hash=sha256:81dfc9e7ffd49f5006685a243905cee72216303e5ea42f6588793dfb8c8407ab \
    --hash=sha256:f3c2b5b691ddf8cabc804f1a7e5c23979634a755f2c363b9449f9808ad43072f
    # via esp-coredump
pyparsing==3.0.9 \
    --hash=sha256:2b020ecf7d21b687f219b71ecad3631f644a47f01403fa1d1036b0c6416d70fb \
    --hash=sha256:5026bae9a10eeaefb61dab2f09052b9f4307d44aee4eda64b309723d8d206bbc
    # via -r requirements.core.in
pyserial==3.5 \
    --hash=sha256:3c77e014170dfffbd816e6ffc205e9842efb10be9f58ec16d3e8675b4925cddb \
    --hash=sha256:c4451db6ba391ca6ca299fb3ec7bae67a5c55dde170964c7a14ceefec02f2cf0
    # via
    #   -r requirements.core.in
    #   esptool
pyyaml==6.0 \
    --hash=sha256:0283c35a6a9fbf047493e3a0ce8d79ef5030852c51e9d911a27badfde0605293 \
    --hash=sha256:055d937d65826939cb044fc8c9b08889e8c743fdc6a32b33e2390f66013e449b \
    --hash=sha256:07751360502caac1c067a8132d150cf3d61339af5691fe9e87803040dbc5db57 \
    --hash=sha256:0b4624f379dab24d3725ffde76559cff63d9ec94e1736b556dacdfebe5ab6d4b \
    --hash=sha256:0ce82d761c532fe4ec3f87fc45688bdd3a4c1dc5e0b4a19814b9009a29baefd4 \
    --hash=sha256:1e4747bc279b4f613a09eb64bba2ba602d8a6664c6ce6396a4d0cd413a50ce07 \
    --hash=sha256:213c60cd50106436cc818accf5baa1aba61c0189ff610f64f4a3e8c6726218ba \
    --hash=sha256:231710d57adfd809ef5d34183b8ed1eeae3f76459c18fb4a0b373ad56bedcdd9 \
    --hash=sha256:277a0ef2981ca40581a47093e9e2d13b3f1fbbeffae064c1d21bfceba2030287 \
    --hash=sha256:2cd5df3de48857ed0544b34e2d40e9fac445930039f3cfe4bcc592a1f836d513 \
    --hash=sha256:40527857252b61eacd1d9af500c3337ba8deb8fc298940291486c465c8b46ec0 \
    --hash=sha256:473f9edb243cb1935ab5a084eb238d842fb8f404ed2193a915d1784b5a6b5fc0 \
    --hash=sha256:48c346915c114f5fdb3ead70312bd042a953a8ce5c7106d5bfb1a5254e47da92 \
    --hash=sha256:50602afada6d6cbfad699b0c7bb50d5ccffa7e46a3d738092afddc1f9758427f \
    --hash=sha256:68fb519c14306fec9720a2a5b45bc9f0c8d1b9c72adf45c37baedfcd949c35a2 \
    --hash=sha256:77f396e6ef4c73fdc33a9157446466f1cff553d979bd00ecb64385760c6babdc \
    --hash=sha256:819b3830a1543db06c4d4b865e70ded25be52a2e0631ccd2f6a47a2822f2fd7c \
    --hash=sha256:897b80890765f037df3403d22bab41627ca8811ae55e9a722fd0392850ec4d86 \
    --hash=sha256:98c4d36e99714e55cfbaaee6dd5badbc9a1ec339ebfc3b1f52e293aee6bb71a4 \
    --hash=sha256:9df7ed3b3d2e0ecfe09e14741b857df43adb5a3ddadc919a2d94fbdf78fea53c \
    --hash=sha256:9fa600030013c4de8165339db93d182b9431076eb98eb40ee068700c9c813e34 \
    --hash=sha256:a80a78046a72361de73f8f395f1f1e49f956c6be882eed58505a15f3e430962b \
    --hash=sha256:b3d267842bf12586ba6c734f89d1f5b871df0273157918b0ccefa29deb05c21c \
    --hash=sha256:b5b9eccad747aabaaffbc6064800670f0c297e52c12754eb1d976c57e4f74dcb \
    --hash=sha256:c5687b8d43cf58545ade1fe3e055f70eac7a5a1a0bf42824308d868289a95737 \
    --hash=sha256:cba8c411ef271aa037d7357a2bc8f9ee8b58b9965831d9e51baf703280dc73d3 \
    --hash=sha256:d15a181d1ecd0d4270dc32edb46f7cb7733c7c508857278d3d378d14d606db2d \
    --hash=sha256:d4db7c7aef085872ef65a8fd7d6d09a14ae91f691dec3e87ee5ee0539d516f53 \
    --hash=sha256:d4eccecf9adf6fbcc6861a38015c2a64f38b9d94838ac1810a9023a0609e1b78 \
    --hash=sha256:d67d839ede4ed1b28a4e8909735fc992a923cdb84e618544973d7dfc71540803 \
    --hash=sha256:daf496c58a8c52083df09b80c860005194014c3698698d1a57cbcfa182142a3a \
    --hash=sha256:e61ceaab6f49fb8bdfaa0f92c4b57bcfbea54c09277b1b4f7ac376bfb7a7c174 \
    --hash=sha256:f84fbc98b019fef2ee9a1cb3ce93e3187a6df0b2538a651bfb890254ba9f90b5
    # via idf-component-manager
reedsolo==1.5.4 \
    --hash=sha256:0f5c33c5abcc51808012aa55f40d06bd15273bae2e54ad9879e41d77a73887ed \
    --hash=sha256:b8b25cdc83478ccb06361a0e8fadc27b376a3dfabbb1dc6bb583a998a22c0127
    # via esptool
requests==2.28.1 \
    --hash=sha256:7c5599b102feddaa661c826c56ab4fee28bfd17f5abca1ebbe3e7f19d7c97983 \
    --hash=sha256:8fefa2a1a1365bf5520aac41836fbee479da67864514bdb821f31ce07ce65349
    # via
    #   idf-component-manager
    #   requests-toolbelt
requests-toolbelt==0.9.1 \
    --hash=sha256:380606e1d10dc85c3bd47bf5a6095f815ec007be7a8b69c878507068df059e6f \
    --hash=sha256:968089d4584ad4ad7c171454f0a5c6dac23971e9472521ea3b6d49d610aa6fc0
    # via idf-component-manager
schema==0.7.5 \
    --hash=sha256:f06717112c61895cabc4707752b88716e8420a8819d71404501e114f91043197 \
    --hash=sha256:f3ffdeeada09ec34bf40d7d79996d9f7175db93b7a5065de0faa7f41083c1e6c
    # via idf-component-manager
six==1.16.0 \
    --hash=sha256:1e61c37477a1626458e36f7b1d82aa5c9b094fa4802892072e49de9c60c4c926 \
    --hash=sha256:8abb2f1d86890a2dfb989f9a77cfcfd3e47c2a354b01111771326f8aa26e0254
    # via
    #   ecdsa
    #   idf-component-manager
tqdm==4.64.0 \
    --hash=sha256:40be55d30e200777a307a7585aee69e4eabb46b4ec6a4b4a5f2d9f11e7d5408d \
    --hash=sha256:74a2cdefe14d11442cedf3ba4e21a3b84ff9a2dbdc6cfae2c34addb2a14a5ea6
    # via idf-component-manager
urllib3==1.26.9 \
    --hash=sha256:44ece4d53fb1706f667c9bd1c648f5469a2ec925fcf3a776667042d645472c14 \
    --hash=sha256:aabaf16477806a5e1dd19aa41f8c2b7950dd3c746362d7e3223dbe6de6ac448e
    # via requests

# The following packages are considered to be unsafe in a requirements file:
setuptools==62.6.0 \
    --hash=sha256:990a4f7861b31532871ab72331e755b5f14efbe52d336ea7f6118144dd478741 \
    --hash=sha256:c1848f654aea2e3526d17fc3ce6aeaa5e7e24e66e645b5be2171f3f6b4e5a178
    # via -r requirements.core.in
```

If this file were committed along with the corresponding .in file, it would be safe from new versions breaking something, and with the hashes, also safe from software supply chain attacks. (Probably not a common problem, just a bonus you get from the additional check)

It would generate something else in the future when idf component manager 1.1.4 is released, or urllib3 releases 1.26.10, etc.

See also the first paragraph of jwhitlock's answer here https://stackoverflow.com/a/36848206/2619499

Otherwise, I understand and will work around it. (In that case, I think your proposed workaround could be included in the official Espressif Dockerfile to prevent future Docker builds from running into this)

I would be interested to know more about the option to skip the constraints file. Is there any chance the option could be made available to export.sh? Perhaps that is what belongs in the Docker image now.

I assume it would be modifying this line in export.sh...

```diff
-python "${IDF_PATH}/tools/idf_tools.py" check-python-dependencies || return 1
+python "${IDF_PATH}/tools/idf_tools.py" check-python-dependencies --no-constraints || return 1
```

or some way that passes it along as a command line switch (does export.sh have any other switches? could just pass the args straight through if not, then `. path/to/export.sh --no-constraints` just works)
